interfaces/builtin/polkit.go

```diff
@@ -145,11 +152,182 @@ func (iface *polkitInterface) PolkitConnectedPlug(spec *polkit.Specification, pl
+	var exptectedHash string
+	for _, installRule := range installRules {
+		if installRule.Name == base {
+			exptectedHash = installRule.Sha3_384
+			break
+		}
+	}
+	if exptectedHash == "" {
+		return nil, fmt.Errorf(`no matching "install-rule" entry found for %q`, filename)
+	}
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	// Check rule file size
+	finfo, err := f.Stat()
+	if err != nil {
+		return nil, err
+	}
+	fsize := finfo.Size()
+	if fsize > maxPolkitRuleFileSize {
+		return nil, fmt.Errorf(`%q is %d bytes, max rule file size is %d`, filename, fsize, maxPolkitRuleFileSize)
+	}
+	// Read rule and compute sha3-384 hash of matched file
+	h := crypto.SHA3_384.New()
+	content := &bytes.Buffer{}
+	w := io.MultiWriter(content, h)
+	if _, err := io.Copy(w, f); err != nil {
+		return nil, fmt.Errorf("cannot obtain hash of %q: %w", filename, err)
+	}
+	hash, err := asserts.EncodeDigest(crypto.SHA3_384, h.Sum(nil))
+	if err != nil {
+		return nil, err
+	}
+	if hash != exptectedHash {
+		return nil, fmt.Errorf("unexpected hash digest of %q, expected %q, found %q", filename, exptectedHash, hash)
+	}
+	// Hash matched, return rule content
+	return polkit.Rule(content.Bytes()), nil
+}
+
+func (iface *polkitInterface) addPolkitRules(spec *polkit.Specification, plug *interfaces.ConnectedPlug) error {
+	installRules, err := iface.parseAndValidateInstallRules(plug)
+	if err != nil {
+		return err
+	}
+
+	mountDir := plug.Snap().MountDir()
+	ruleFiles, err := filepath.Glob(filepath.Join(mountDir, "meta", "polkit", plug.Name()+".*.rules"))
+	if err != nil {
+		return err
+	}
+	if len(ruleFiles) == 0 {
+		return fmt.Errorf("cannot find any rule files for plug %q", plug.Name())
+	}
+	for _, filename := range ruleFiles {
+		suffix := strings.TrimSuffix(filepath.Base(filename), ".rules")
+		rule, err := readPolkitRule(filename, installRules)
+		if err != nil {
+			return err
+		}
+		if err := spec.AddRule(suffix, rule); err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+type polkitMissingAttrErr struct {
+	snapName string
+}
+
+func (err *polkitMissingAttrErr) Error() string {
+	return fmt.Sprintf(`snap %q must have at least one of ("action-prefix", "install-rules") attributes set for interface "polkit"`, err.snapName)
+}
+
+func (iface *polkitInterface) PolkitConnectedPlug(spec *polkit.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// At least one of ("action-prefix", "install-rules") attributes must be set.
+	policyErr := iface.addPolkitPolicies(spec, plug)
+	if policyErr != nil && !errors.Is(policyErr, snap.AttributeNotFoundError{}) {
+		return policyErr
+	}
+	ruleErr := iface.addPolkitRules(spec, plug)
+	if ruleErr != nil && !errors.Is(ruleErr, snap.AttributeNotFoundError{}) {
+		return ruleErr
+	}
+	// Check if both attributes are not set.
+	if policyErr != nil && ruleErr != nil {
+		return &polkitMissingAttrErr{plug.Snap().InstanceName()}
+	}
```

Comment:
wouldn't this be caught already in BeforePreparePlug() and so any errors we observe here should be returned?

Reply:
We can't return errors here because it could only be `AttributeNotFoundError` which is ignored above and other errors would have already been returned, This only checks that at least one of the attributes is declared which is redundant so maybe we could simply remove this check?

Reply:
I think so, similarly BeforePreparePlug should have caught and reacted on AttributeNotFoundError, so there's no need to check for it in the `if` braches above